src/extractor/graph_compressor.cpp

```diff
@@ -327,6 +330,19 @@ void GraphCompressor::Compress(const std::unordered_set<NodeID> &barrier_nodes,
                                reverse_node_weight_penalty,
                                reverse_node_duration_penalty);
 
+                auto add_expanded_traffic_signal =
+                    [&expanded_traffic_signals,
```

Comment:
The change to traffic signals needs to be made in-place, as they're subsequently used as part of the edge-expanded graph generation too. As mentioned in the [other thread](https://github.com/Project-OSRM/osrm-backend/pull/6153#issuecomment-1788694243), we could make a compressor wrapper for traffic signals, much like we do with TurnPaths.

https://github.com/Project-OSRM/osrm-backend/blob/31e31a63d062fb804f5f4695ed3036ca7a269ead/src/extractor/turn_path_compressor.cpp#L36

Reply:
> The change to traffic signals needs to be made in-place, as they're subsequently used as part of the edge-expanded graph generation too. As mentioned in the [other thread](https://github.com/Project-OSRM/osrm-backend/pull/6153#issuecomment-1788694243), we could make a compressor wrapper for traffic signals, much like we do with TurnPaths.
> 
> https://github.com/Project-OSRM/osrm-backend/blob/31e31a63d062fb804f5f4695ed3036ca7a269ead/src/extractor/turn_path_compressor.cpp#L36

Thank you very much for your guidance :smile:. I will try to make a compressor wrapper for traffic signals, like you do with TurnPaths. I need some time to complete the modification, and add test cases to the Cucumber suite.

Reply:
Hi @mjjbell, I'm sorry that i only started working on this recently. I study the **TurnPathCompressor** class, but i find that the traffic signal problem seems to be different.

--- Related method ---
https://github.com/Project-OSRM/osrm-backend/blob/31e31a63d062fb804f5f4695ed3036ca7a269ead/src/extractor/edge_based_graph_factory.cpp#L622-L624
https://github.com/Project-OSRM/osrm-backend/blob/31e31a63d062fb804f5f4695ed3036ca7a269ead/src/extractor/edge_based_graph_factory.cpp#L630
The **TrafficSignals** info used during the generation of the edge-expanded graph. But the **HasSignal** method is not affected by [src/extractor/graph_compressor.cpp](https://github.com/Project-OSRM/osrm-backend/blob/31e31a63d062fb804f5f4695ed3036ca7a269ead/src/extractor/graph_compressor.cpp). As the **GetLastEdgeSourceID** method returns the original node_id of the segment (even if the node has been compressed), the _from_node_ & _intersection_node_ are the original relation in the OSM nodes. So we don't have to worry about the **HasSignal** losing traffic signal feature due to **_compression_**.

Do we still need to make a compressor wrapper for traffic signals?

Reply:
Yes, well, this is currently working only **because** we're not considering the compression here.

If we start compressing the traffic signals, this becomes simpler, as we will not longer care if it's a trivial edge or not.

For example, this currently works as expected

```
@routing @car @traffic_light
Feature: Car - Handle traffic lights

    Background:
        Given the profile "car"

    Scenario: Car - Traffic signal direction turn with edge compression
        Given the node map
            """
                    d
                    |
                    2
                    |
            a-1-b - c--f
                    |
                    e

            """

        And the ways
            | nodes | highway |
            | abc   | primary |
            | ce    | primary |
            | cd    | primary |
            | ce    | primary |

        And the nodes
            | node | highway         | traffic_signals:direction | # |
            | c    | traffic_signals | forward                   | ambiguous, but happens in OSM

        When I route I should get
            | from | to | time   | weight | #                             |
            | 1    | 2  |  35.1s | 35.1   | turn with traffic light    |
            | 2    | 1  |  29.9s | 29.9   | turn with no traffic light |
```

But it would fail if we correctly compress the traffic signals and don't update the edge based graph factory logic.

The case that is currently not being handled properly is when there is no turn, as shown here
https://github.com/Project-OSRM/osrm-backend/pull/6153#issuecomment-2001979000

Reply:
ok, do you still need me to make a compressor wrapper for traffic signals ？

Reply:
No, I included it in the traffic signal, as it's simple enough.
https://github.com/Project-OSRM/osrm-backend/blob/8526cc7d45adce9c352897acfd0366211e4f1d5e/include/extractor/traffic_signals.hpp#L23